add codecov.io support

tylerjw opened the pull request:
This adds codecov.io report uploading support.  As noted in the README.md change in order to use these changes a cmake INTERFACE library needs to be added to the build instructions to enable coverage report generation on only your project's files.  It might work to instead add these options through CMAKE_ARGS but then it'll generate a coverage report for all of the code in your project (including tests).  There is a way to prune these with a codecov.yaml file.  Let me know if you would rather me implement it this way.

The copying of the build directory out of docker allows the codecov.io script to find the artifacts it needs to compile the report.  I tried running the codecov.io script inside of the docker image but that didn't work because it couldn't reach codecov.io.  I don't understand networking in docker images enough to troubleshoot this.

Thank you for reviewing this.

henningkayser commented on README.md:
```diff
@@ -143,3 +148,39 @@ It's also possible to run the script without using docker. To this end, issue th
+
+## Enabling codecov.io reporting
+
+To enable codecov enabling you'll need to add this to your CMakeLists.txt file:
```
"enable codecov enabling"?

henningkayser commented on README.md:
```diff
@@ -143,3 +148,39 @@ It's also possible to run the script without using docker. To this end, issue th
+      target_link_libraries(coverage_config INTERFACE --coverage)
+    endif()
+
+Then you'll need to change the `target_link_libraries` command to turn on code coverage specifically for your project without propigating these settings to external dependencies:
```
```suggestion
Then you'll need to change the `target_link_libraries` command to turn on code coverage specifically for your project without propagating these settings to external dependencies:
```

henningkayser commented on after_success.sh:
```diff
@@ -0,0 +1,16 @@
+
+if [[ "${TEST:=}" == *code-coverage* ]]; then
+  echo -e $(colorize BOLD "Generating codecov.io report")
+  travis_run --title "codecov.io report upload" bash <(curl -s https://codecov.io/bash)
```
This is almost too easy ;)

tylerjw replied:
It was too easy, I discovered that reports weren't actually being generated because the build directory was still owned by root and not the travis user.  I'm updating this with one more command to fix this.

tylerjw commented:
> You mean "curl -s https://codecov.io/bash" doesn't work for querying the script? That's odd, we use wget and curl for different purposes all the time within docker. Uploading a report is not possible from within a nested docker container right away. Passing the environment variables to the docker run command doesn't work? https://docs.codecov.io/docs/testing-with-docker.

Curl works just fine.  It is the uploading that fails.  It fails because it can't reach codecov.io to send the report.

tylerjw commented on README.md:
```diff
@@ -143,3 +148,39 @@ It's also possible to run the script without using docker. To this end, issue th
+    )
+
+    target_link_libraries(
+      ${PROJECT_NAME}
```
It builds them as part of the PUBLIC link policy above.  This was to separate what gets built with the --coverage flag.  I'm going to do some more testing and see if I can simplify this.

tylerjw commented:
@rhaschke Thank you for the review.  I am responding to give you an update on this and respond to your specific concerns and get your feedback on the approach I'm taking.

> 1. that you need to copy the build folder. See Henning's comment [#91 (review)](https://github.com/ros-planning/moveit_ci/pull/91#pullrequestreview-309808798) for a possible solution.

This works, I've got this working on my repo I'm using for testing and verifying this change and will include this in my next commit for this.

> 2. that we need to augment the project's cmake files. If you add the `--coverage` option globally, I guess a coverage report is generated for all packages. Manually listing them in codecov.yaml for pruning is error-prone as well. But you could just delete all corresponding files not belonging to packages that are actually tested. We do similar things for [blacklisting those packages for unit testing](https://github.com/ros-planning/moveit_ci/blob/840f25fe1a81557e18d06ca69c45712019530a19/travis.sh#L264-L268).

I haven't found a way to do this without editing cmake files, however I have found a more [robust solution](https://github.com/Lectem/cpp-boilerplate).  In that repo there is a `Coverage.cmake` file that can be included with `include(Coverage)`.  This checks for compiler support before changing the compiler flags and allows you to enable it with `-DCMAKE_BUILD_TYPE=Coverage"`.  Then you have to exclude folders and files using the `codecov.yaml` file.  However the syntax for it is nice enough:
```
ignore:
  - "test"
  - "external"
  - "*_main.cpp"
```

What are your thoughts on this revised approach?  Once I get it working on my test repo I will push a commit with these changes.

rhaschke commented:
Dear Tyler,

I think I now better understand how coverage works: Code that shouldn't be considered for coverage (testing source code and external stuff), shouldn't be compiled with instrumentation flags. Hence, your initial approach to define an interface library target that can be "linked" with appropriated build items, seems to be most reasonable.
Obviously we cannot really avoid touching the cmake files. If you want to keep the Coverage.cmake file, for MoveIt it might be useful to provide that file once in moveit_core and re-use it across all the other packages.

tylerjw commented:
Hello Robert,

I just pushed an updated version of this.  I'm now going to now attempt to get this working with the moveit project to verify my approach.  After experimenting with both I think the `.codecov.yml` file exclude list is easier to understand (and therefore easier for people who maintain/contribute to a project to use) than any approach I derived using cmake file changes to exclude files.  Please don't spend time reviewing these changes until I've validated it working with moveit as I might need to change my approach again here to make the changes to moveit to be as simple as possible.

Thank you,
Tyler

rhaschke commented on travis.sh:
```diff
@@ -33,6 +33,16 @@ function run_script() {
+    # these two varaibles are needed for codecov to work locally
+    set +u  # stop variable checking
+    # test if they are set before setting as travis sets them
+    [[ -z ${VCS_BRANCH_NAME} ]] && export VCS_BRANCH_NAME=$(git rev-parse --abbrev-ref HEAD)
```
This is needed for local (non-Travis) usage, I guess?

tylerjw replied:
yes

rhaschke replied:
Could you please get rid off `set +u` / `set -u` by using corresponding defaults?

tylerjw replied:
What do you mean by corresponding defaults, is that a different way of testing if a variable is set?

rhaschke replied:
I implemented my suggestion in https://github.com/PickNikRobotics/moveit_ci/pull/1.
Having this feature added, you should also augment the unit tests accordingly.

tylerjw replied:
I added unit tests for this.

rhaschke commented on travis.sh:
```diff
@@ -33,6 +33,16 @@ function run_script() {
 
+    # get ci enviroment parameters to pass into docker for codecov
+    # this is a list of docker parameters to include enviroment variables
+    export CI_ENV_PARAMS=`bash <(curl -s https://codecov.io/env)`
```
Remove `export` here?

rhaschke commented on travis.sh:
```diff
@@ -287,9 +295,10 @@ function test_workspace() {
 
-   # Copy build products for code coverage reporting
-   [[ "${TEST:=}" == *code-coverage* ]] && travis_run_simple cp -r $ROS_WS/build /root/$REPOSITORY_NAME
+function send_codecov_report() {
```
This is a one-line function that is only used once. Please move content to the call location.

tylerjw commented:
@rhaschke Today I started trying to get this to work for moveit.  Locally I can never get `catkin build run_tests` to finish.  I pushed this up to see if it would finish and travis fails at the same point that my laptop does.  It always times out with this message:

`The command "catkin build --catkin-make-args run_tests -- --no-status --summarize 2>/dev/null" reached the internal timeout of 20 minutes. Aborting.`

The one thing I tried was building with optimization.  All of the examples I've found of people using this build with `-O0 -g` and that is set in that cmake file I linked.  I tried removing this so it would build with `-O3`.  That helped it get further but it still had the issue of timing out while running integration tests.

Do you have any ideas on how I could get the integration test to not timeout?  Is there some tests I should disable?  If I did that though the code coverage number would be artificially too low.  The only other idea I have is to do the cmake changes the way I originally suggested.  The downside to this is it will be a ton of changes to adjust build options per source file.  Let me know if you can think of a simpler way.

tylerjw commented:
I tested this working over on moveit_grasps.  Here is an example of the sort of report it generates: https://codecov.io/gh/ros-planning/moveit_grasps/branch/p%2Ftylerjw%2Fcodecov

I will try to figure out how to add the unit test to this next.

tylerjw commented:
Taking this out of WIP as I think it is ready again for a review.  I added to the unit test for it and have verified it works on moveit_grasps.

davetcoleman commented on README.md:
```diff
@@ -143,3 +144,25 @@ It's also possible to run the script without using docker. To this end, issue th
+
+## Enabling codecov.io reporting
+
+For codecov to work you need to build and link your c++ code with specific parameters.  To enable this we use the ros package [code_coverage](https://github.com/mikeferguson/code_coverage).  To to use the `code-coverage` test in your repo make these two changes:
```
```suggestion
For codecov to work you need to build and link your C++ code with specific parameters.  To enable this we use the ROS package [code_coverage](https://github.com/mikeferguson/code_coverage).  To to use the `code-coverage` test in your repo make these two changes:
```

davetcoleman commented:
> I'm wondering what this would imply for the current effort to get CircleCi+Codecov running for MoveIt 2: ros-planning/moveit2#44

@ruffsl 

Also, having @mikeferguson look at this would be ideal as he made the original package.

tylerjw commented:
I'm testing adding support for TEST_WHITELIST which will use the catkin config --whitelist command before calling test to split up running different packages' tests across different travis runs.  I'm doing this to see if I can solve the issue with the moveit repo timing out on running code coverage for all tests since it takes much longer to run with code coverage turned on.

tylerjw commented:
I figured out an issue for the timeouts on moveit :)  However master on moveit is broken so I can't fully verify.  I think this change is ready to be merged now.

tylerjw commented on travis.sh:
```diff
@@ -290,6 +302,12 @@ function test_workspace() {
+   TEST_WHITELIST=$(unify_list " ,;" ${TEST_WHITELIST:-})
+   echo -e $(colorize YELLOW Test whitelist: $(colorize THIN $TEST_WHITELIST))
+   test -n "$TEST_WHITELIST" && catkin config --whitelist $TEST_WHITELIST &> /dev/null
+
```
With these new lines I was able to split up the code coverage run on moveit over two travis runs and solve the timeout issue.  This is what I put int he moveit `.travis.yml` file:

```
- TEST="code-coverage" TEST_WHITELIST="moveit_core"
- TEST="code-coverage" TEST_BLACKLIST="moveit_ros_perception moveit_core"
```

rhaschke commented on README.md:
```diff
@@ -143,3 +144,25 @@ It's also possible to run the script without using docker. To this end, issue th
+
+## Enabling codecov.io reporting
+
+For codecov to work you need to build and link your C++ code with specific parameters.  To enable this we use the ROS package [code_coverage](https://github.com/mikeferguson/code_coverage).  To to use the `code-coverage` test in your repo make these two changes:
```
```suggestion
For codecov to work you need to build and link your C++ code with specific parameters.  To enable this we use the ROS package [code_coverage](https://github.com/mikeferguson/code_coverage).  Using the `code-coverage` test in your repo requires the following two changes:
```

rhaschke commented on README.md:
```diff
@@ -143,3 +144,25 @@ It's also possible to run the script without using docker. To this end, issue th
+For codecov to work you need to build and link your C++ code with specific parameters.  To enable this we use the ROS package [code_coverage](https://github.com/mikeferguson/code_coverage).  To to use the `code-coverage` test in your repo make these two changes:
+
+1. Add `<test_depend>code_coverage</test_depend>` to your package.xml
+1. Add this to your `CMakeLists.txt`:
```
```suggestion
2. Add this to your `CMakeLists.txt`:
```

rhaschke commented on test_pkgs/valid/CMakeLists.txt:
```diff
@@ -10,6 +10,13 @@ include_directories(
 
-if(CATKIN_ENABLE_TESTING)
+if (CATKIN_ENABLE_TESTING)
+  find_package(code_coverage REQUIRED)
```
Shouldn't this moved into the scope of the next `if` block?

rhaschke commented on README.md:
````diff
@@ -143,3 +144,25 @@ It's also possible to run the script without using docker. To this end, issue th
+
+```cmake
+# to run: catkin_make -DENABLE_COVERAGE_TESTING=ON package_name_coverage
+if(CATKIN_ENABLE_TESTING AND ENABLE_COVERAGE_TESTING)
````
Maybe split up this if condition into two separate ones to match your code snippet below?

tylerjw replied:
This is the exact code that is in the moveit codebase already.  I just copied it from there.  What I did in the unit test is the minimal amount needed.  What is kind of nice about this if statement (like what is in the moveit codebase) is it adds the codecoverage to one section with out being two blocks of code like what's in the example from code_coverage ros package's README.md.

tylerjw replied:
This is coppied code from the cmake files in moveit.  I like how it keeps the code coverage section separate from the testing if section.  I can change it if you'd perfer.  I made the unit test match this.

rhaschke commented on README.md:
```diff
@@ -143,3 +144,25 @@ It's also possible to run the script without using docker. To this end, issue th
+  include(CodeCoverage)
+  APPEND_COVERAGE_COMPILER_FLAGS()
+  set(COVERAGE_EXCLUDES "*/test/*")
+  add_code_coverage(NAME ${PROJECT_NAME}_coverage)
```
Is this really needed? You didn't use this command below for the unit test.

tylerjw replied:
Yes, I'll add it to the unit test.

rhaschke commented on test_pkgs/valid/CMakeLists.txt:
```diff
@@ -11,12 +11,13 @@ include_directories(
+	find_package(code_coverage REQUIRED)   # catkin package ros-*-code-coverage
+  include(CodeCoverage)
+  APPEND_COVERAGE_COMPILER_FLAGS()
+  add_code_coverage(NAME ${PROJECT_NAME}_coverage)
```
`add_code_coverage` is not required. This just adds a new make target to create the reports manually.

rhaschke commented on test_pkgs/valid/CMakeLists.txt:
```diff
@@ -11,12 +11,13 @@ include_directories(
+
+# to run: catkin_make -DENABLE_COVERAGE_TESTING=ON package_name_coverage
+if(CATKIN_ENABLE_TESTING AND ENABLE_COVERAGE_TESTING)
+	find_package(code_coverage REQUIRED)   # catkin package ros-*-code-coverage
```
Please fix indentation.

tylerjw replied:
good catch, fixed

rhaschke commented on README.md:
```diff
@@ -143,3 +144,25 @@ It's also possible to run the script without using docker. To this end, issue th
+  include(CodeCoverage)
+  APPEND_COVERAGE_COMPILER_FLAGS()
+  set(COVERAGE_EXCLUDES "*/test/*")
+  add_code_coverage(NAME ${PROJECT_NAME}_coverage)
```
This line is not needed.

tylerjw replied:
removed

tylerjw commented:
I originally dismissed the approach of using the code already in the moveit cmake files because I didn't understand it and I thought it was just for generating the html output.

tylerjw commented:
This worked for the first time on the moveit library.  One thing you'll notice is that this is only turned on for the moveit_core directory as that is the only one with support for it in cmake.

https://codecov.io/gh/ros-planning/moveit/pull/1719/tree

tylerjw commented:
Another thing to note is this does not generate reports for tests that test python code.  I will create an issue to add that as a feature in the future.

rhaschke commented on README.md:
```diff
@@ -159,7 +159,6 @@ if(CATKIN_ENABLE_TESTING AND ENABLE_COVERAGE_TESTING)
   include(CodeCoverage)
   APPEND_COVERAGE_COMPILER_FLAGS()
   set(COVERAGE_EXCLUDES "*/test/*")
```
I'm pretty sure that the `COVERAGE_EXCLUDES` variable has no effect here. This is only relevant for use with the coverage report target created by `add_code_coverage()`. But this we don't use here, because we use the bash script provided by codecov.io.
Did you validated the effect of `COVERAGE_EXCLUDES`?

tylerjw replied:
You are right :(.  If you look at the files that were "tested" in the above linked codecov.io report you will find tests.  Good catch.  I think we'll need to bring back the codecov yaml file.  I'll test that now and update this.

tylerjw commented:
@rhaschke I believe I've addressed all your feedback on this.  Can we submit this and I'll start a new PR if I find I need to make additional changes as I integrate this with moveit?

rhaschke commented on test_pkgs/valid/CMakeLists.txt:
```diff
@@ -10,6 +10,13 @@ include_directories(
 )
 
-if(CATKIN_ENABLE_TESTING)
+if (CATKIN_ENABLE_TESTING)
```
```suggestion
if(CATKIN_ENABLE_TESTING)
```

rhaschke commented on test_pkgs/valid/CMakeLists.txt:
```diff
@@ -10,6 +10,13 @@ include_directories(
   catkin_add_gtest(${PROJECT_NAME} test.cpp)
 endif()
+
+# to run: catkin_make -DENABLE_COVERAGE_TESTING=ON package_name_coverage
```
```suggestion
```

rhaschke commented:
Before merging this, I would like to see that tests are actually ignored. In https://github.com/ros-planning/moveit/pull/1719, I see that you are still struggling with this. Is it possible that coverage reports are accumulated, so you don't see if tests are ignored now when they were not ignored in the past?

tylerjw commented:
@rhaschke I'll let you know when I get that working.  Hopefully very soon.

mikeferguson commented on README.md:
```diff
@@ -143,3 +144,30 @@ It's also possible to run the script without using docker. To this end, issue th
+# to run: catkin_make -DENABLE_COVERAGE_TESTING=ON package_name_coverage
+if(CATKIN_ENABLE_TESTING AND ENABLE_COVERAGE_TESTING)
+  find_package(code_coverage REQUIRED)   # catkin package ros-*-code-coverage
+  include(CodeCoverage)
```
I just released 0.3.0 with @rhaschke improvements - once that propagates through the build farm and gets sync'd you'll be able to drop this line

tylerjw replied:
Which line?  The find_package or the include?  Can you link me to your changes?

mikeferguson replied:
The include -- see https://github.com/mikeferguson/code_coverage/pull/15 for details (again, I just did the release, it's probably 2 weeks before it syncs all the way through the build farm).

tylerjw replied:
Ah, cool.  Hopefully I'll get this merged quickly and then I can make a small PR to remove this line once your change propagates.

tylerjw commented:
I don't think I understand why, but you don't have to run the tests to do code coverage reporting.  I'm going to update this to reflect that and it will speed up the code-coverage builds considerably.

tylerjw commented:
I just got it to work!  @rhaschke checkout my latest changes.  I couldn't figure out a way to exclude all test directories through the codecov.yml however digging through the options of the bash uploader for codecov.io I found an argument that is then passed to a find instruction before it complies the report.  With this change any path that contains a directory called `test` is excluded from the reporting.

rhaschke commented on travis.sh:
```diff
@@ -93,6 +93,8 @@ function run_docker() {
         -e CXX=${CXX_FOR_BUILD:-${CXX:-c++}} \
         -e CFLAGS \
         -e CXXFLAGS \
+        -e CCACHE_SLOPPINESS \
```
[ccache has a default cache size of 5GB](https://ccache.dev/manual/latest.html#_cache_size_management). Did you check whether these extra options have a measurable effect? If not, I suggest to drop them to reduce complexity.

tylerjw replied:
It was only using 500mb because the default size on travis is 500mb.

rhaschke commented on travis.sh:
```diff
@@ -282,8 +284,18 @@ function build_workspace() {
-   # For a command that doesn’t produce output for more than 10 minutes, prefix it with travis_run_wait
-   travis_run_wait 60 --title "catkin build" catkin build --no-status --summarize
+   # Set compile flag to improve ccache hit rate
+   export CXXFLAGS="$CXXFLAGS -fdebug-prefix-map=$(pwd)=."
```
I'm curious: How does this effect ccache? Isn't this a debugging option of gcc?

tylerjw replied:
That idea came from this: https://cppisland.wordpress.com/2018/12/30/ccache-quick-guide/

The basics is that with a Debug build there is a chance the hit rate is low because of paths in the debug symbols and this is one way to fix that.  This should have no affect on non-debug builds.

rhaschke replied:
Thanks for providing the link. However, their use case is different: They are building the same stuff in _different_ folders, while for us all files reside in the very same folder between build attempts. Hence, there shouldn't be a cache miss due to cwd.
Did you look into the log files generated with export `CCACHE_DEBUG=1` between to consecutive (local) builds?

rhaschke commented on travis.sh:
```diff
@@ -282,8 +284,18 @@ function build_workspace() {
+   echo "CXXFLAGS = $CXXFLAGS"
+
+   # If test whitelist is set, explicitly build that project
+   if [ "${TEST_WHITELIST:-}" ]; then
```
Why so complicated? If `TEST_WHITELIST` is empty, the if-branch command equals the else-branch one.

tylerjw replied:
Good point, I didn't think of that.

tylerjw commented:
@rhaschke The reason I'm adding ccache changes to this PR is because I think debug symbols are the reason I'm getting so bad of cache performance and that is a big part of why the builds are so slow.  On the plus side here is the first full (assuming I added the cmake changes to everywhere I should have) coverage report for moveit:

https://codecov.io/gh/PickNikRobotics/moveit/tree/70809b9ac08d5baf6ddeed521622e79ae3e680d4

tylerjw commented:
Some other things I want to try to see if they work better to improve ccache:

Increase cache size to 20G (not sure the impact on travis cache storage).
Set CCACHE_BASEDIR=$(pwd) instead of that compiler debug flag

tylerjw commented:
See my PR against @rhaschke 's other codecov PR.
